File: src/com/ptby/dynamicreturntypeplugin/scanner/FunctionCallReturnTypeScanner.java
package com.ptby.dynamicreturntypeplugin.scanner;

import com.jetbrains.php.lang.psi.elements.impl.FunctionReferenceImpl;
import com.ptby.dynamicreturntypeplugin.config.FunctionCallConfig;
import com.ptby.dynamicreturntypeplugin.gettype.GetTypeResponse;
import com.ptby.dynamicreturntypeplugin.typecalculation.CallReturnTypeCalculator;

import java.util.List;

public class FunctionCallReturnTypeScanner {

    private final CallReturnTypeCalculator callReturnTypeCalculator;


    public FunctionCallReturnTypeScanner( CallReturnTypeCalculator callReturnTypeCalculator ) {

        this.callReturnTypeCalculator = callReturnTypeCalculator;
    }


    public GetTypeResponse getTypeFromFunctionCall( List<FunctionCallConfig> functionCallConfigs,
                                                    FunctionReferenceImpl functionReference ) {
        for ( FunctionCallConfig functionCallConfig : functionCallConfigs ) {
            if ( functionCallConfig.equalsFunctionReference( functionReference ) ) {
                GetTypeResponse getTypeResponse = callReturnTypeCalculator
                        .calculateTypeFromFunctionParameter( functionReference, functionCallConfig.getParameterIndex()
                        );
                if ( !getTypeResponse.isNull() && functionCallConfig.hasValidStringClassNameMask() ) {
                    String maskReplacedType = String.format(
                            functionCallConfig.getStringClassNameMask(), getTypeResponse.toString()
                    );

                    return new GetTypeResponse( maskReplacedType );
                }

                return getTypeResponse;
            }
        }

        return new GetTypeResponse( null );
    }
}

File: src/com/ptby/dynamicreturntypeplugin/scanner/FunctionCallReturnTypeScanner.kt
package com.ptby.dynamicreturntypeplugin.scanner

import com.jetbrains.php.lang.psi.elements.FunctionReference
import com.ptby.dynamicreturntypeplugin.config.FunctionCallConfigKt
import com.ptby.dynamicreturntypeplugin.gettype.FunctionReferenceGetTypeResponse
import com.ptby.dynamicreturntypeplugin.typecalculation.ParameterTypeCalculator

public class FunctionCallReturnTypeScanner() {
    private val parameterTypeCalculator: ParameterTypeCalculator

    init {
        parameterTypeCalculator = ParameterTypeCalculator()
    }

    public fun getTypeFromFunctionCall(functionCallConfigs: List<FunctionCallConfigKt>,
                                       functionReference: FunctionReference): FunctionReferenceGetTypeResponse {
        for (functionCallConfig in functionCallConfigs) {
            if (functionCallConfig.equalsFunctionReference(functionReference)) {
                return FunctionReferenceGetTypeResponse.newFunction(functionReference)
            }
        }

        return FunctionReferenceGetTypeResponse.createNull()
    }

}
